File: server/src/follow/follow.service.ts
import { BadRequestException, Injectable, NotFoundException } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { NotificationService } from '../notification/notification.service';

@Injectable()
export class FollowService {
  constructor(
    private prisma: PrismaService,
    private notificationService: NotificationService,
  ) {}

  async followUser(followerId: string, followingId: string) {
    if (followerId === followingId) throw new BadRequestException('Cannot follow yourself');

    const user = await this.prisma.user.findUnique({ where: { id: followingId } });
    if (!user) throw new NotFoundException('User not found');

    const existing = await this.prisma.follow.findUnique({
      where: { followerId_followingId: { followerId, followingId } },
    });

    if (existing) {
      await this.prisma.follow.delete({
        where: { followerId_followingId: { followerId, followingId } },
      });
      return { following: false };
    }

    await this.prisma.follow.create({ data: { followerId, followingId } });
    await this.notificationService.createNotification({
      userId: followingId,
      actorId: followerId,
      type: 'FOLLOW',
    });
    return { following: true };
  }

  async getFollowers(userId: string) {
    const user = await this.prisma.user.findUnique({ where: { id: userId } });
    if (!user) throw new NotFoundException('User not found');

    return this.prisma.follow.findMany({
      where: { followingId: userId },
      include: {
        follower: { select: { id: true, name: true, profileSrc: true } },
      },
    });
  }

  async getFollowing(userId: string) {
    const user = await this.prisma.user.findUnique({ where: { id: userId } });
    if (!user) throw new NotFoundException('User not found');

    return this.prisma.follow.findMany({
      where: { followerId: userId },
      include: {
        following: { select: { id: true, name: true, profileSrc: true } },
      },
    });
  }
}

File: client/lib/services/follow_service.dart
import 'package:client/services/api_client.dart';

class FollowService {
  Future<bool> isFollowing(String userId) async {
    final response = await dio.get('/api/follow/$userId/is-following');
    return response.data['isFollowing'];
  }

  Future<bool> followUser(String userId) async {
    final response = await dio.post('/api/follow/$userId');
    return response.data['following'];
  }

  Future<List<Map<String, dynamic>>> getFollowers(String userId) async {
    final response = await dio.get('/api/follow/$userId/followers');
    final List data = response.data;
    return data.map((item) => {
      'id': item['follower']['id'],
      'name': item['follower']['name'],
      'image': item['follower']['profileSrc'],
    }).toList().cast<Map<String, dynamic>>();
  }

  Future<List<Map<String, dynamic>>> getFollowing(String userId) async {
    final response = await dio.get('/api/follow/$userId/following');
    final List data = response.data;
    return data.map((item) => {
      'id': item['following']['id'],
      'name': item['following']['name'],
      'image': item['following']['profileSrc'],
    }).toList().cast<Map<String, dynamic>>();
  }
}

final followService = FollowService();